File: 백준/Silver/9020. 골드바흐의 추측/골드바흐의 추측.py
import sys
input = sys.stdin.readline

# 에라토스테네스의 체
sosu = [True] * 10001
sosu[0], sosu[1] = False, False

for i in range(2, 10001):
    for j in range(i*2, 10001, i):
        if sosu[j]: sosu[j] = False

t = int(input())
for i in range(t):
    n = int(input())
    ans = []
    for i in range(n+1):
        if sosu[i] and sosu[n-i]:
            ans.append((i, n-i))
    
    diff = 10000
    ans1, ans2 = 0, 0
    for a, b in ans:
        if diff > abs(a-b):
            ans1 = a
            ans2 = b
            diff = abs(a-b)
    print(ans1, ans2)

File: 백준/Silver/9020. 골드바흐의 추측/골드바흐의 추측.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int abs(int first, int second)
{
	if (first >= second) return (first - second);
	return (second - first);
}

int main()
{
	int T, n;
	cin >> T;

	int sosu[10001] = {0};
	sosu[0] = 1;
	sosu[1] = 1;

	// 에라토스테네스의 체
	// 값이 0인 경우가 소수임
	for (int i = 2; i < 10001; i++) {
		for (int j = i*2; j < 10001; j += i) {
			if (sosu[j] == 0) sosu[j] = 1;
		}
	}

	for (int i = 0; i < T; i++) {
		cin >> n;
		vector<pair<int, int>> ans;

		// 쌍 저장
		for (int j = 0; j < n; j++) {
			if (sosu[j] == 0 && sosu[n-j] == 0) {
				ans.push_back(make_pair(j, n-j));
			}
		}

		// 정답 탐색
		int diff = 10000;
		pair<int, int> answer;
		for (auto p : ans) {
			if (diff > abs(p.first, p.second)) {
				answer = p;
				diff = abs(p.first, p.second);
			}
		}

		cout << answer.first << " " << answer.second << endl;
	}

	return 0;
}